kep/624-disk-io-aware-scheduling/README.md

```diff
@@ -0,0 +1,207 @@
+
+## Design Details
+
+The design includes the following key components: the disk IO scheduler plugin and its interfaces with the IO Driver.
+
+<p align="center"><img src="images/key_components.png" title="Key components" width="600" class="center"/></p>
+The disk IO scheduler plugin communicates with the IO Driver to pass information on IO metrics collection context (e.g., reserved pod list), obtain a normalized IO BW for each new pod’s IO BW request, and retrieve updates on each disk’s normalized available IO capacity for making scheduling decisions.
+
+The IO Driver, which is to be implemented by disk IO vendors, comprises three components.
+
+### IO Metrics Collector
+
+The IO Metrics Collector, which runs on each worker node, acts as an IO metric collector and analyzer. It watches the actual disk IO utilization of each pod, calculates the disk’s available IO capacity based on each workload’s real-time usage and characteristics using the IO Calculation Model and reports the node’s real-time available disk IO capacity to the aggregator when it is smaller than what is saved in the scheduler plugin’s cache. Since different disk vendors could have different ways of collecting metric, this component is outside the scope of the scheduler plugin’s implementation.
+
+### Aggregator
+
+The aggregator consolidates the IO metrics from each worker node, reports a consolidated list of real-time available disk IO capacity to the scheduler plugin. It is also responsible for converting each new pod’s disk IO BW request to a normalized value using the disk’s IO Calculation Model so as to match the disk’s normalized available IO capacity. Since the normalization function is disk type and vendor specific, this component is outside the scope of the scheduler plugin’s implementation.
+
+### IO Calculation Model
+
+The IO Calculation Model is responsible for converting the disk’s available IO capacity and each new pod’s IO request to normalized values. Since the normalization function is disk type and vendor specific, this component is outside the scope of the scheduler plugin’s implementation.
+
+### Disk IO Scheduler Plugin
+
+We leverage the K8s scheduler framework to add the disk IO scheduler plugin.
+When the scheduler plugin starts, it subscribes to the IO Driver with a long streaming connection. The IO Driver updates the scheduler plugin about each node disk’s normalized available IO capacity and the scheduler plugin stores the info in the plugin’s local cache. We thought about using CRD mechanism before. But going through the CRD route, it has two drawbacks: 
+1. In the pod scheduling flow, the disk IO scheduler plugin needs to pass the disk IO info specified in the annotation of the pod spec to the vendor-specific IO driver and get back a normalized disk IO BW value. Using CRD to support this bi-directional communication would involve creating two CRDs and two watches, which introduces long latency to the pod scheduling flow. 
```

Comment:
The reason that the scheduler plugin needs to communicate with the IO driver is as follows: the disk IO BW specified in the workload's pod spec is a value when the workload performs the disk IO operation using the workload's associated block size and read/write ratio. Different workloads could have different block sizes and read/write ratios. It is known that a workload performing the same disk IO operation but using different block sizes or different read/write ratios would consume different disk IO BW. Therefore to be consistent, we need to normalize all the pods' disk IO BW requirements. Because the normalizing parameters are disk vendor-dependent, the scheduler plugin needs to interface with a vendor-specific driver to get the normalized IO BW value. The communication is gRPC-based and should be fast given the small amount of msg communication.

Reply:
I think you are right. We will investigate this path. Thanks!

Reply:
@Huang-Wei Thanks for the suggestion. This path works. We will update the KEP to have the scheduler plugin get the disk-specific normalization function from the IO driver upon startup time and cache it. Then at the pod scheduling cycle, the scheduler plugin will invoke the normalization function directly without talking to the gRPC server.